lib/auth0/mixins/validation.rb

```diff
@@ -20,6 +28,118 @@ def validate_permissions_array(permissions)
       end
 
+      class IdTokenValidator
+        def initialize(context)
+          @context = context
+        end
+
+        def validate(id_token)
+          begin
+            result = JWT.decode id_token, nil, false, {}
+          rescue JWT::DecodeError
+            raise Auth0::InvalidIdToken, 'ID token could not be decoded'
+          end
+
+          decoded_id_token = { payload: result.first, header: result.last, string: id_token }
+          validate_signature decoded_id_token
+          validate_claims decoded_id_token
+        end
+
+        private
+
+        def validate_signature(decoded_id_token)
+          algorithm = @context[:algorithm]
+          algorithm_name = algorithm.class.to_s.split('::').last
+          options = { algorithms: [algorithm_name], verify_expiration: false, verify_not_before: false }
+          secret = nil
+
+          case algorithm
+          when Auth0::Algorithm::RS256
+            options[:jwks] = JSON.parse(JSON[algorithm.jwks], symbolize_names: true)
+          when Auth0::Algorithm::HS256
+            secret = algorithm.secret
+          end
+
+          begin
+            JWT.decode decoded_id_token[:string], secret, true, options
+          rescue JWT::VerificationError
+            raise Auth0::InvalidIdToken, 'Invalid ID token signature'
+          rescue JWT::IncorrectAlgorithm
+            alg = decoded_id_token[:header]['alg']
+            raise Auth0::InvalidIdToken, "Signature algorithm of \"#{alg}\" is not supported. Expected the ID token to be signed with \"#{algorithm_name}\""
+          rescue JWT::DecodeError
+            kid = decoded_id_token[:header]['kid']
+            raise Auth0::InvalidIdToken, "Could not find a public key for Key ID (kid) \"#{kid}\""
+          end
+        end
+
+        def validate_claims(decoded_id_token); end
+      end
+
+      module Algorithm
+        # Represents the HS256 algorithm, which rely on shared secrets.
+        # @see https://auth0.com/docs/tokens/concepts/signing-algorithms
+        class HS256
+          class << self
+            private :new
+
+            # Create a new instance passing the shared secret.
+            # @param secret [string] The HMAC shared secret.
+            # @return [object] A new instance.
+            def secret(secret)
+              new secret
+            end
+          end
+
+          attr_accessor :secret
+
+          def initialize(secret)
+            raise Auth0::InvalidParameter, 'Must supply a valid secret' if secret.to_s.empty?
+
+            @secret = secret
+          end
+        end
+
+        # Represents the RS256 algorithm, which rely on public key certificates.
+        # @see https://auth0.com/docs/tokens/concepts/signing-algorithms
+        class RS256
+          include Auth0::Mixins::HTTPProxy
+
+          @@cache = Zache.new.freeze
+
+          class << self
+            private :new
+
+            # Create a new instance passing the shared secret.
```

Comment:
these docs are for the HS256 case

Reply:
Fixed in efdf3feec312ed9d2e8a91fa3ea9b625373033ec